fix access to files that located inside a JAR.

kfriedberger opened the pull request:
This has failed on Windows 10 with OpenJDK 15 before.
See #34 for details.

kfriedberger committed:
fix access to files that located inside a JAR.

This has failed on Windows 10 with OpenJDK 15 before.
See #34 for details.

kfriedberger committed:
Close filesystem after opening to avoid a resource-leak.

This commit restructures parts of the method to align
with the scopes of filesystem and path handling.

PhilippWendler commented on src/org/sosy_lab/common/configuration/ConfigurationBuilder.java:
```diff
@@ -260,9 +262,26 @@ public ConfigurationBuilder loadFromResource(Class<?> contextClass, String resou
+          }
+        }
+      }
       sourcePath = Paths.get(url.toURI());
```
```suggestion
      sourcePath = Paths.get(uri);
```

PhilippWendler commented on src/org/sosy_lab/common/configuration/ConfigurationBuilder.java:
```diff
@@ -260,9 +262,26 @@ public ConfigurationBuilder loadFromResource(Class<?> contextClass, String resou
     @Var Path sourcePath;
     @Var String sourceString;
     try {
+      // try to access the requested file.
```
This does not try to access the given file, so please rewrite the comment.

Given that the whole code is quite non-obvious as it relies on the fact that `FileSystemProvider.newFileSystem` has some side effect, this really needs good documentation, with an explanation of the problem that this solves, why it solves it, and a link to the source of the JAR.

kfriedberger replied:
I restructured the code and added some documentation.
Please ask if anything remains unclear.

PhilippWendler replied:
Well, this comment is still as wrong as before, right?

PhilippWendler commented on src/org/sosy_lab/common/configuration/ConfigurationBuilder.java:
```diff
@@ -260,9 +262,26 @@ public ConfigurationBuilder loadFromResource(Class<?> contextClass, String resou
+        for (FileSystemProvider provider : FileSystemProvider.installedProviders()) {
+          if (provider.getScheme().equalsIgnoreCase("jar")) {
+            try {
+              provider.getFileSystem(uri);
```
Calling `getFileSystem` and throwing away its result is confusing, as this method should not have a side effect. Is the only reason why we call this method to determine whether the file system already exists?

In this case, I would find the following code much clearer:
```java
// Ensure file system for uri exists
try {
  provider.newFileSystem(uri, ImmutableMap.of());
} catch (FileSystemNotFoundException ignored) {
  // already exists
}
```

kfriedberger replied:
Opening a new filesystem registers a FileSystemProvider, thus the URI can be resolved.
This seems to be some background magic of Java.

I agree with your code change and will change it.

PhilippWendler reviewed (changes requested):
Did you test whether this actually solves #34? This is not obvious for me. After all, the exception from #34 is not the one thrown in the `try` block that is modified here, where a `Path` is created from a `URI`, but in a different method, where we try to construct the same `Path` from a string-

This solution has the side effect that a new file system is registered and stays in the list of file systems permanently, i.e., this is a (small) memory leak. Is there a way to avoid this?

Furthermore, this seems to open the referenced JAR file, although we will never read from it. Could this be avoided?

kfriedberger committed:
fix Java version for Refaster.

PhilippWendler commented on src/org/sosy_lab/common/configuration/ConfigurationBuilder.java:
```diff
@@ -250,35 +253,72 @@ public ConfigurationBuilder loadFromFile(Path file)
    * @throws IllegalArgumentException If the resource cannot be found or read, or contains invalid
    *     syntax or #include directives.
    */
+  @SuppressWarnings("try")
```
Why is this necessary?

kfriedberger replied:
Because the FileSystem in the try-with-ressources-block is never used locally.

PhilippWendler commented on src/org/sosy_lab/common/configuration/ConfigurationBuilder.java:
```diff
@@ -250,35 +253,72 @@ public ConfigurationBuilder loadFromFile(Path file)
-    } catch (URISyntaxException | FileSystemNotFoundException | IllegalArgumentException e) {
+      // try to access the requested file.
+      URI uri = url.toURI();
+      try (FileSystem fs = getFileSystemForUri(uri)) {
```
This closes the `FileSystem` afterwards, but the documentation of `FileSystemProvider.newFileSystem` says

> Once a file system is closed it is provider-dependent if the provider allows a new file system to be created with the same URI as a file system it previously created.

So this could mean that calling this method twice fails, right? Or is it guaranteed for the provider for JAR files that it will succeed to reopen the file twice? Is there any documentation on this?

Furthermore, if the file system already exists before this method is called, won't this even mean that we close it although somebody else has registered it?

kfriedberger replied:
Yes, in general it depends on the FileSystems whether it can be reopened.
For JAR files, a ZipFileSystem is used and it seems to support multiple openings and closings.

Parallel access might be a problem: we could close a first instance and an interleaved concurrent reading process is then aborted. There is still not a solution for that.

I will change the code to not close, when already opened from other callee.

PhilippWendler commented on src/org/sosy_lab/common/configuration/ConfigurationBuilder.java:
```diff
@@ -250,35 +253,72 @@ public ConfigurationBuilder loadFromFile(Path file)
 
-    return this;
+  /**
+   * If the URI is part of a JAR file, we open the file system from the JAR. If this fails, we
```
If this is just relevant for JARs, its name should reflect this.

PhilippWendler commented on src/org/sosy_lab/common/configuration/ConfigurationBuilder.java:
```diff
@@ -250,35 +253,72 @@ public ConfigurationBuilder loadFromFile(Path file)
       // This will not allow resolving relative path names, but everything else works.
-      sourcePath = null;
-      sourceString = url.toString();
+      parseSource(contextClass, resourceName, source, Optional.empty(), url.toString());
```
Won't this still also the same kind of exceptions if we pass a string as source that was not able to be converted to a `Path` in this method?

PhilippWendler replied:
This was not addressed? I see no way how this could be expected to work when `Paths.get()` is called on the result of `url.toString()` later, given that the conversion via `URI` has already failed?

kfriedberger replied:
From my point, all cases that resulted in an execution of that branch, are handled as before, except the case, where a JAR file is accessed.
Thus, if it failed before, it will also fail now.
Currently I have no test case where I run into this branch.
Could you give an example why this case is required?

Do we have any JUnit test for this code?

PhilippWendler replied:
Sure, the current code improves the situation, but now that we recognize that other cases will still fail for the same reason, we should fix these, too, right?

We cannot unit test this properly because it requires interfacing with Java APIs and use of URLs.

kfriedberger replied:
Right, I thought the same: testing might not be possible.
Which further cases do you know that should be handled?
Can we write a documentation for manual testing at least?

PhilippWendler replied:
Actually, the following code might work for testing:
```
    URL url =
        new URL(
            "https://raw.githubusercontent.com/sosy-lab/cpachecker/trunk/config/includes/resource-limits.properties");
    CharSource source = Resources.asCharSource(url, StandardCharsets.UTF_8);
    Configuration.builder().loadFromSource(source, url.toString(), url.toString());
```
This uses `loadFromSource`, but this in turn calls the same `parse` method as `loadFromResource`, it is just easier to specify the parameters. Note that the last parameter is `toString()` of an URL, just like in the `catch` block of `loadFromResource`.

For me, this creates a `UnixPath` with the `https://...`, which happens to work. But I guess it would crash on Windows?

kfriedberger replied:
Yes, `https://...` crashes on Windows.
And I could not find any fix, because there is no matching FileSystemProvider available.
By default, `FileSystemProvider.installedProviders()` only lists the providers for `[UnixFileSystemProvider, JrtFileSystemProvider]` or `[WindowsFileSystemProvider, JrtFileSystemProvider]`, depending on the operating system. Any further filesystem access, e.g., accessing zip-files or http-files on Linux, might not be specified as default by Java itself, but provided by the environment.

Thus, I would not expect further support from Java and stay with the minimal working solution, which is already implemented in this PR.

PhilippWendler replied:
Well, the fix could be to not rely on `Path` if it does not support what we need (as mentioned in https://github.com/sosy-lab/java-common-lib/issues/34#issuecomment-728745129).

This is somewhat tricky, because these `Path`s end up in public API, but at least we can avoid the crashes now and refactor some things to make all this more clear. I just pushed a commit that does this, please review (and ideally, test again on Windows).

PhilippWendler reviewed (dismissed):
You did not address my question about how this is related to solving #34, and also some other questions?

The added comments also does not explain the relevant part, because the exception in #34 is not from calling `Paths.get` inside `loadFromResource`. And there is no explanation why we need the file system to be open if `Path` uses that internally anyway and where this code come from.

Please either address or respond to all of my comments, please do not just ignore some of them.

kfriedberger committed:
ignore the FS if is is already open, and more documentation.

kfriedberger committed:
formatting

kfriedberger commented:
The implemented solution works so far, e.g., the cmdline from #34 works.
The exception reported in #34 is just an after-effect from the real problem,
thus, the modified try-block appears some lines before the actual problematic part,
e.g., where the first exception (`FileSystemNotFoundException`) was swallowed.
The changed code that works fine, because the file system for the JAR file is open before access.

kfriedberger committed:
fix typo in exception.

PhilippWendler committed:
Improve handling of source information in configuration parser

So far, we required a string, and this string needed to be a valid path.
Otherwise, the parser would always crash.
However, the only known user of this information (except in error
messages) was FileTypeConverter, which already gracefully handled the
case of missing source information.

Furthermore, sometimes we also used dummy source strings to cover the
case that no source information is present, but this was only checked in
one case and not for example by FileTypeConverter.

So now we officially make "missing source" a supported case of our
Configuration class, remove the dummy string, and avoid the parser crash
if source is not a valid Path when loading configs from URLs.

PhilippWendler committed:
Syntactically change default value of rootDirectory for FileTypeConverter

Due to the previous commit, rootDirectory is now used more often
(whenever no source info is present, so far it often used a dummy value).
With "." as rootDirectory we could lots of changes (e.g., in tests)
from "file" to "./file", which is equivalent.
By changing the rootDirectory from "." to the equivalent "",
we change those cases where the default value of the rootDirectory
was previously used, but these should be fewer cases.

And anyway, now it is more consistent.

PhilippWendler committed:
Restore previous behavior of FileTypeConverter

So far, FileTypeConverter used the rootDirectory only for default values
of options, because for all non-default values the Configuration
instance actually passed a non-null source.
This differed from the documented behavior.
The second-to-last commit changed this, but for example some tests in
CPAchecker relied on the current behavior.
We can actually restore the previous behavior without difficulty,
so let's do this. The behavior is now not really consistent and
different than what was documented, but this could be improved
separately if desired.

PhilippWendler committed:
apply refaster suggestion
